src/storage/src/source/mysql/replication.rs

```diff
@@ -110,84 +118,113 @@ pub(crate) fn render<G: Scope<Timestamp = TransactionId>>(
 
             // Get the set of GTIDs currently available in the binlogs (GTIDs that we can safely start replication from)
             let binlog_gtid_set: String = conn.query_first("SELECT GTID_SUBTRACT(@@global.gtid_executed, @@global.gtid_purged)").await?.unwrap();
-            // NOTE: The assumption here for the MVP is that there is only one source-id in the GTID set, so we can just
-            // take the transaction_id from the first one. Fix this once we move to a partitioned timestamp.
-            let binlog_start_gtid = GtidSet::from_str(&binlog_gtid_set)?.first().expect("At least one gtid").to_owned();
-            let binlog_start_transaction_id = TransactionId::new(binlog_start_gtid.earliest_transaction_id());
-
+            // For each GTID UUID in the binlog set, get a partitioned timestamp representing the transaction-id of the first transaction that is available
+            let binlog_start_gtids: BTreeMap<Uuid, _> = match gtid_singletons(GtidSet::from_str(&binlog_gtid_set)?, false) {
+                Ok(res) => res,
+                // If GTID intervals in the binlog are not available in a monotonic consecutive order this breaks all of our assumptions
+                // and there is nothing else we can do. This can occur if the mysql server is restored to a previous point-in-time
+                // or if a user manually adds transactions to the @@gtid_purged system var.
+                Err(err) => return Ok(
+                    return_definite_error(err, &output_indexes, &mut data_output, data_cap_set, &mut definite_error_handle, definite_error_cap_set).await
+                )
+            }.into_iter().map(|part| (part.interval().singleton().unwrap().clone(), part)).collect();
+            trace!(%id, "timely-{worker_id} replication binlog contains gtids starting at: {binlog_start_gtids:?}");
+
+            // Calculate the resume upper that represents the start transaction-id for each GTID UUID
+            // that we should resume from, across all sub-sources.
             let resume_upper = Antichain::from_iter(
                 subsource_resume_uppers
                     .values()
                     .flat_map(|f| f.elements())
-                    // Advance any upper as far as the start of the binlog.
-                    .map(|t| std::cmp::max(*t, binlog_start_transaction_id))
+                    // Remove any partitions representing UUIDs that do not yet exist
+                    .filter(|t| t.timestamp() != &TransactionId::Absent)
+                    // Advance each partition's upper as far as the start transaction-id of the corresponding GTID UUID available in the binlog
```

Comment:
I suppose this is here because pg is doing it too, but this is a bug in pg that needs to be fixed. We shouldn't fast forward the transaction ids here because that risks losing data. Instead, what we want is to calculate the minimum frontier across all subsources which is simply `Antichain::from_iter(subsource_resume_uppers.into_values().flatten())`